src/socket.c

```diff
@@ -0,0 +1,221 @@
+	hints.ai_flags = AI_PASSIVE;
+
+	*slist = NULL;
+
+	/* listen to all available (IPv4 and IPv6) address */
+	if ((r = getaddrinfo(NULL, portname, &hints, &res)) != 0) {
+		fprintf(stderr, "getaddrinfo() failed: %s\n", gai_strerror(r));
+		return -1;
+	}
+
+	for (ptr = res; ptr != NULL; ptr = ptr->ai_next) {
+		s = socket(ptr->ai_family, SOCK_STREAM, 0);
+		if (s < 0) {
+			perror("socket() failed");
+			continue;
+		}
+
+		if (ptr->ai_family == AF_INET)
+			fprintf(stderr, "Listening on %s:%d\n", inet_ntop(ptr->ai_family,
+				&((struct sockaddr_in*)ptr->ai_addr)->sin_addr, strip,
+				sizeof(strip)), port);
+		else if (ptr->ai_family == AF_INET6)
+			fprintf(stderr, "Listening on [%s]:%d\n", inet_ntop(ptr->ai_family,
+				&((struct sockaddr_in6*)ptr->ai_addr)->sin6_addr, strip,
+				sizeof(strip)), port);
+
+#if defined(IPV6_V6ONLY)
+		if (ptr->ai_family == AF_INET6) {
+			y = 1;
+			/* avoid listen on ipv6 addresses failing
+			 * because already listening on ipv4 addresses: */
+			setsockopt(s, IPPROTO_IPV6, IPV6_V6ONLY,
+				   (const void *) &y, sizeof(y));
+		}
+#endif
+
+		y = 1;
+		if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR,
+			       (const void *) &y, sizeof(y)) < 0) {
+			perror("setsockopt(SO_REUSEADDR) failed");
+		}
+
+		if (bind(s, ptr->ai_addr, ptr->ai_addrlen) < 0) {
+			perror("bind() failed");
+			close(s);
+			continue;
+		}
+
+		if (listen(s, 1024) < 0) {
+			perror("listen() failed");
+			close(s);
+			r = -1;
+			goto cleanup;
+		}
+
+		lm = calloc(1, sizeof(socket_list));
+		if (lm == NULL) {
+			close(s);
+			r = -1;
+			goto cleanup;
+		}
+		lm->s = s;
+		lm->family = ptr->ai_family;
+		lm->addrlen = ptr->ai_addrlen;
+		memcpy(&lm->addr, ptr->ai_addr, ptr->ai_addrlen);
+		lm->next = *slist;
+		*slist = lm;
+	}
+
+	if (*slist == NULL)
+		r = -1;
+	else
+		r = 0;
+
+ cleanup:
+	freeaddrinfo(res);
+	fflush(stderr);
+
+	return r;
+}
+
+static void spawn_process(int fd, const char *jwkdir,
+						  process_request_func pfunc,
+						  socket_list *slist)
+{
+	pid_t pid;
+	socket_list *ptr;
+
+	pid = fork();
+	if (pid == 0) { /* child */
+		for(ptr=slist; ptr != NULL; ptr=ptr->next) {
+			close(ptr->s);
+		}
+		/* Ensure that both stdout and stdin are set */
+		if (dup2(fd, STDOUT_FILENO) < 0) {
+			perror("dup2");
+			close(fd);
+			return;
+		}
+
+		close(fd);
+
+		pfunc(jwkdir, STDOUT_FILENO);
+		exit(0);
+	} else if (pid == -1) {
+		perror("fork failed");
+	}
+	close(fd);
+}
+
+static void handle_child(int sig)
+{
+	pid_t pid;
+	int status;
+
+	while ((pid = waitpid(-1, &status, WNOHANG)) > 0);
+}
+
+int run_service(const char *jwkdir, int port, process_request_func pfunc)
+{
+	socket_list *slist, *ptr;
+	int r, n = 0, accept_fd;
+	fd_set read_fds;
+	struct timeval tv;
+
+	signal(SIGCHLD, handle_child);
+
+	r = listen_port(&slist, port);
+	if (r < 0) {
+		fprintf(stderr, "Could not listen port (%d)\n", port);
+		return -1;
+	}
+
+	while(1) {
+		FD_ZERO(&read_fds);
+		for(ptr=slist; ptr != NULL; ptr=ptr->next) {
```

Comment:
This should be homogeneous to other assignations:
for (ptr = slist; ptr != NULL; ptr = ptr->next) ....